rhp/v4/rpc.go

```diff
@@ -0,0 +1,718 @@
+	defer s.Close()
+
+	req := rhp4.RPCFormContractRequest{
+		Prices:        p,
+		Contract:      params,
+		Basis:         basis,
+		MinerFee:      formationTxn.MinerFee,
+		RenterInputs:  renterSiacoinElements,
+		RenterParents: formationSet,
+	}
+	if err := rhp4.WriteRequest(s, rhp4.RPCFormContractID, &req); err != nil {
+		return RPCFormContractResult{}, fmt.Errorf("failed to write request: %w", err)
+	}
+
+	var hostInputsResp rhp4.RPCFormContractResponse
+	if err := rhp4.ReadResponse(s, &hostInputsResp); err != nil {
+		return RPCFormContractResult{}, fmt.Errorf("failed to read host inputs response: %w", err)
+	}
+
+	// add the host inputs to the transaction
+	var hostInputSum types.Currency
+	for _, si := range hostInputsResp.HostInputs {
+		hostInputSum = hostInputSum.Add(si.Parent.SiacoinOutput.Value)
+		formationTxn.SiacoinInputs = append(formationTxn.SiacoinInputs, si)
+	}
+
+	if n := hostInputSum.Cmp(fc.TotalCollateral); n < 0 {
+		return RPCFormContractResult{}, fmt.Errorf("expected host to fund at least %v, got %v", fc.TotalCollateral, hostInputSum)
+	} else if n > 0 {
+		// add change output
+		formationTxn.SiacoinOutputs = append(formationTxn.SiacoinOutputs, types.SiacoinOutput{
+			Address: fc.HostOutput.Address,
+			Value:   hostInputSum.Sub(fc.TotalCollateral),
+		})
+	}
+
+	// sign the renter inputs
+	signer.SignV2Inputs(&formationTxn, toSign)
+	formationSigHash := cs.ContractSigHash(fc)
+	formationTxn.FileContracts[0].RenterSignature = signer.SignHash(formationSigHash)
+
+	renterPolicyResp := rhp4.RPCFormContractSecondResponse{
+		RenterContractSignature: formationTxn.FileContracts[0].RenterSignature,
+	}
+	for _, si := range formationTxn.SiacoinInputs[:len(renterSiacoinElements)] {
+		renterPolicyResp.RenterSatisfiedPolicies = append(renterPolicyResp.RenterSatisfiedPolicies, si.SatisfiedPolicy)
+	}
+	// send the renter signatures
+	if err := rhp4.WriteResponse(s, &renterPolicyResp); err != nil {
+		return RPCFormContractResult{}, fmt.Errorf("failed to write signature response: %w", err)
+	}
+
+	// read the finalized transaction set
+	var hostTransactionSetResp rhp4.RPCFormContractThirdResponse
+	if err := rhp4.ReadResponse(s, &hostTransactionSetResp); err != nil {
+		return RPCFormContractResult{}, fmt.Errorf("failed to read final response: %w", err)
+	}
+
+	if len(hostTransactionSetResp.TransactionSet) == 0 {
+		return RPCFormContractResult{}, fmt.Errorf("expected at least one host transaction")
+	}
+	hostFormationTxn := hostTransactionSetResp.TransactionSet[len(hostTransactionSetResp.TransactionSet)-1]
+	if len(hostFormationTxn.FileContracts) != 1 {
+		return RPCFormContractResult{}, fmt.Errorf("expected exactly one contract")
+	}
+
+	// check for no funny business
+	formationTxnID := formationTxn.ID()
+	hostFormationTxnID := hostFormationTxn.ID()
+	if formationTxnID != hostFormationTxnID {
+		return RPCFormContractResult{}, fmt.Errorf("expected transaction IDs to match %v != %v", formationTxnID, hostFormationTxnID)
```

Comment:
in my experience there's little benefit to printing mismatched hashes; unlike other comparisons, it rarely tells you anything beyond "they are different"